File: scripts/placement_bot.py
#!/usr/bin/env python3
"""
Placement Bot for Ugandan Internship & Industrial Training Intelligence
Setup: pip install -r ../requirements.txt
"""

import time
import requests
from bs4 import BeautifulSoup
from urllib.parse import quote_plus

SECTORS = [
    "Agribusiness & Forestry",
    "Healthcare & Medical",
    "Media & ICT",
    "Finance & Commerce",
    "Tourism & Hospitality",
    "Engineering & Technical",
    "Legal & Professional Services",
]

PLACEMENT_TYPES = ["industrial training", "undergraduate internship", "graduate trainee"]

EXCLUSIONS = [".ac.ug", ".edu", "wikipedia.org", "curriculum", "syllabus", "admissions", "tuition"]
STALE_YEARS = ["2023", "2022", "2021", "2020", "2019"]
USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36"
)


def clean_links(links: list[str]) -> list[str]:
    cleaned = []
    for link in links:
        lower = link.lower()
        if any(bad in lower for bad in EXCLUSIONS):
            continue
        if any(year in lower for year in STALE_YEARS):
            continue
        cleaned.append(link)
    return cleaned


def google_search(query: str) -> list[str]:
    url = f"https://www.google.com/search?q={quote_plus(query)}"
    headers = {"User-Agent": USER_AGENT}
    res = requests.get(url, headers=headers, timeout=10)
    if res.status_code != 200:
        return []
    soup = BeautifulSoup(res.text, "html.parser")
    links = []
    for g in soup.select(".tF2Cxc"):
        a = g.select_one("a")
        if a and a.get("href"):
            links.append(a["href"])
    return clean_links(links)


def build_queries(sector: str, placement_type: str, region: str | None = None) -> list[str]:
    base = f'Uganda "{sector}" {placement_type} 2025'
    region_part = f" {region}" if region else ""
    return [
        f"{base}{region_part}",
        f"{base}{region_part}",  # Global search across all domains
        f"{base}{region_part} site:(.ug OR .com OR .org OR .net OR .dev OR .co.ug OR .ac.ug)",
        f"{base}{region_part} (site:org OR site:gov OR site:org.ug OR site:go.ug)",
        f"{sector} graduate trainee Uganda 2025 -job-boards",
        f"{base} (Kampala OR Entebbe OR Mbarara OR Jinja)",
    ]


def scrape_uganda_placements(sectors: list[str], region: str | None = None):
    all_hits: list[dict] = []
    for sector in sectors:
        for placement_type in PLACEMENT_TYPES:
            queries = build_queries(sector, placement_type, region)
            for q in queries:
                print(f"[SCAN] {q}")
                links = google_search(q)
                for link in links:
                    all_hits.append({
                        "sector": sector,
                        "placement_type": placement_type,
                        "query": q,
                        "url": link,
                    })
                time.sleep(20)  # throttle to avoid detection
    return all_hits


def main():
    results = scrape_uganda_placements(SECTORS)
    print("\n--- RESULTS ---")
    for r in results:
        print(f"[{r['sector']} | {r['placement_type']}] {r['url']}")


if __name__ == "__main__":
    main()

File: src/lib/placementBot.ts
import type { PlacementFilters, PlacementQuery, PlacementSector, PlacementType } from '@/types/placements';

const baseUserAgent =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36";

const sectorSynonyms: Record<PlacementSector, string[]> = {
  "Agribusiness & Forestry": ["agriculture", "forestry", "fisheries", "timber"],
  "Healthcare & Medical": ["hospital", "clinic", "medical", "healthcare"],
  "Media & ICT": ["media", "journalism", "ict", "software", "telecom"],
  "Finance & Commerce": ["banking", "finance", "insurance", "audit", "commerce"],
  "Tourism & Hospitality": ["tourism", "hotel", "hospitality", "travel", "lodge"],
  "Engineering & Technical": ["engineering", "construction", "manufacturing", "automotive"],
  "Legal & Professional Services": ["law", "legal", "consultancy", "hr", "advisory"],
};

const exclusionFragments = [".ac.ug", ".edu", "wikipedia.org", "curriculum", "syllabus", "admissions", "tuition"];
const staleYears = ["2023", "2022", "2021", "2020", "2019"];

function encodeQuery(q: string) {
  return encodeURIComponent(q.trim());
}

function buildDork(filters: PlacementFilters) {
  const parts: string[] = [];
  if (filters.keywords) parts.push(filters.keywords);
  if (filters.sector) parts.push(filters.sector);
  if (filters.placementType) parts.push(filters.placementType);
  parts.push("career opportunity");
  parts.push("professional development");
  // Geo bias if provided
  if (filters.region) parts.push(filters.region);
  // Only current/future
  parts.push("2025");
  return parts.filter(Boolean).join(" ");
}

export function buildPlacementQueries(filters: PlacementFilters): PlacementQuery[] {
  const dork = buildDork(filters);
  const sector = filters.sector ?? "";

  const google = `https://www.google.com/search?q=${encodeQuery(dork)}`;
  const googleAllDomains = `https://www.google.com/search?q=${encodeQuery(`${dork} site:(.ug OR .com OR .org OR .net OR .dev OR .co.ug OR .ac.ug)`)}`;
  const googleOrgGov = `https://www.google.com/search?q=${encodeQuery(`${dork} opportunity (site:org.ug OR site:go.ug OR site:org OR site:gov)`)}`;
  const googleProfessional = `https://www.google.com/search?q=${encodeQuery(`${sector} career development Uganda 2025 -job-boards`)}`;
  const googleCities = `https://www.google.com/search?q=${encodeQuery(`${dork} opportunity (Kampala OR Entebbe OR Mbarara OR Jinja)`)}`;

  return [
    { label: "Global search - all domains (.ug, .com, .org, .dev, etc.)", url: google, notes: "Searches across all domain extensions for Uganda-based opportunities." },
    { label: "All Uganda domains (.ug, .co.ug, .ac.ug, .com, .org, .net, .dev)", url: googleAllDomains, notes: "Explicitly targets multiple Ugandan-related domain extensions." },
    { label: "NGO & Govt opportunities", url: googleOrgGov, notes: "Focuses on org/gov domains for public and NGO opportunities worldwide." },
    { label: "Professional development tracks", url: googleProfessional, notes: "Looks for professional growth programs and excludes generic job boards." },
    { label: "City-focused opportunities", url: googleCities, notes: "Bias toward key cities (Kampala, Entebbe, Mbarara, Jinja) for on-the-ground experience." },
  ];
}

export function filterOutNoise(urls: string[]): string[] {
  return urls.filter((u) => {
    const lower = u.toLowerCase();
    if (exclusionFragments.some((frag) => lower.includes(frag))) return false;
    if (staleYears.some((y) => lower.includes(y))) return false;
    return true;
  });
}

export const placementBotConfig = {
  userAgent: baseUserAgent,
  sectorSynonyms,
  exclusionFragments,
  staleYears,
};
